### ISSUE TITLE
Sentry key is null in raven-js when passing SENTRY_DSN only to start process

### ISSUE THREAD

**Issue by damienbeaufils** [nuxt-community/sentry-module#15]
Hello,

I have a Nuxt v1.4.0 application which I build once with `npm run build`, to be able to ship it in a Docker container. In the container, the application is started using `npm start`.
The container is deployed in multiple environments (development, staging, production...) and we use environment variables to have distinct configurations between environments. So I have a distinct `SENTRY_DSN` environment variable between staging and production.

When I start the application with the environment variable, Sentry configuration is OK for node (`raven`) but not for client (`raven-js`). It seems that the `raven-js` client configuration is only filled during build phase, from the `nuxt-config` variables.

Would it be possible to build without that information but to start the application with it?
For example: `npm run build && SENTRY_DSN=xyz npm start`
In other words, would it be possible for the generated code be updated with the environment variable during application start?

If not, I can't build once and run everywhere the application, because I have `SENTRY_DSN` values depending on the environment on which the application is started.

Thanks!

**Comment by DiederikvandenB** [comment]
Unfortunately, I do not believe this is possible. What you do by running `npm run build` is compile the code your code from ES6 in to ugly JS **so that it runs on the client**. For what you want to be possible, each user on your website would have to bring their own environment file containing exactly your settings.

This is why Nuxt's `asyncData` function provides the `env` object in it's `context`, because that is ran serverside. If you'd try to echo environment variables anywhere outside of that function, you will see their value is undefined.

Having said that, you can try to figure out client side what the current environment is (check the url, ip address, etc). It's not the prettiest solution maybe (I unfortunately do not know the best practice in your specific scenario), but it will resolve your issue in the sense that you can build once and run everywhere.

Hope this helps, if not, please reach out!

**Comment by williamchong007** [comment]
@DiederikvandenB
Would it be possible to inject the server ENV DSN via middleware/asyncData, then init client sentry with this DSN on runtime?
The problem I am facing is that I want to use dockerhub for automated webpack build, but currently I have to hardcode my DSN into the my repo, and I need to rebuild a whole new image just for changing the DSN
It would be much simpler if I can just swap the DSN in my server config file, while using a same docker image for deployment.

**Comment by manniL** [comment]
All environment variables (except server settings like PORT, HOST, SOCKET, ...) will be grabbed on build (not on run time/`nuxt-start`) by Nuxt. IIRC there are people who wrote workarounds for that, but it'd be easier to build your app per server (with the different settings) I guess.

**Comment by aldarund** [comment]
@williamchong007 u can try to use https://www.npmjs.com/package/nuxt-env

**Comment by DiederikvandenB** [comment]
@williamchong007 I _think_ that it is possible what you are proposing, but in that case you will only receive client side thrown exceptions. However, as @manniL, there might be workarounds for that too, although I would not necessarily recommend it.

To be honest, I don't see what the big problem is in building separate staging & production docker images.

What you could maybe do is running `yarn run build` on the server (potentially as the entrypoint in your Dockerfile) where you'll be deploying, but in a sense that defeats the purpose of Docker containers.

**Comment by DiederikvandenB** [comment]
> @williamchong007 u can try to use https://www.npmjs.com/package/nuxt-env

Please note this warning in their docs:

> WARNING: As with the config.env option in Nuxt config, environment variables used in nuxt-env are exposed client side, so don't use it for storing secrets!

**Comment by SAGV** [comment]
For anyone who prefers to build the app in advance and wants to pass DSN on start, use https://www.npmjs.com/package/nuxt-env and start client-side separately. 

For this you'd need to add `@sentry/browser` to the `package.json`, preferably, same versions as nuxt-sentry, like `"@sentry/browser": "4.6.6",`. And create a separate client-side sentry plugin which will pick configuration variables via nuxt-env module. In this case, I've just copy-pasted current client plugin and adjusted it:

```
// Taken from the https://github.com/nuxt-community/sentry-module/blob/master/lib/sentry.client.js and adjusted to support node-env
import Vue from 'vue'
import * as Sentry from '@sentry/browser'

export default function (ctx, inject) {
  const opts = Object.assign({}, {
    dsn: ctx.app.$env.SENTRY_DSN
  }, {
    integrations: [
      new Sentry.Integrations.Dedupe(),
      new Sentry.Integrations.ExtraErrorData(),
      new Sentry.Integrations.InboundFilters(),
      new Sentry.Integrations.FunctionToString(),
      new Sentry.Integrations.TryCatch(),
      new Sentry.Integrations.Breadcrumbs(),
      new Sentry.Integrations.GlobalHandlers(),
      new Sentry.Integrations.LinkedErarors(),
      new Sentry.Integrations.UserAgent(),
      new Sentry.Integrations.Vue({ Vue })
    ]
  })

  // Enable Sentry or report to the console that it is not enabled on production
  if (ctx.app.$env.SENTRY_DSN) {
    Sentry.init(opts)
    if (ctx.app.$env.NODE_ENV !== 'production') console.error('Warning! Sentry enabled on non-production env')
  } else if (ctx.app.$env.NODE_ENV === 'production') {
    console.error('Client-side sentry is not setup, DSN is missing')
  }

  // Inject Sentry to the context as $sentry
  ctx.$sentry = Sentry
  inject('sentry', Sentry)
}
```


And adjust your `nuxt.config.js` to run you custom client-side module:

```

plugins: [
    { src: '~/plugins/sentry', ssr: false } // ssr:false disables it when rendered server-side
],
modules: [
    '@nuxtjs/sentry',
    ['nuxt-env', {
      keys: ['SENTRY_DSN', 'NODE_ENV']
    }]
],
sentry: {
    dsn: process.env.SENTRY_DSN,
    disableClientSide: true // Is activated separately via plugins/sentry so it parses sentry DSN via nuxt-env
}
```
